src/scheduler/Scheduler.cpp

```diff
@@ -1129,4 +1162,145 @@ ExecGraphNode Scheduler::getFunctionExecGraphNode(unsigned int messageId)
 
     return node;
 }
+
+bool Scheduler::checkForMigrationOpportunities(
+  faabric::util::MigrationStrategy migrationStrategy)
+{
+    // Vector to cache all migrations we have to do, and update the shared map
+    // at the very end just once. This is because we need a unique lock to write
+    // to the shared map, but the rest of this method can do with a shared lock.
+    std::vector<std::shared_ptr<faabric::PendingMigrations>>
+      tmpPendingMigrations;
+    SPDLOG_INFO("Checking for migration opportunities");
+
+    {
+        faabric::util::SharedLock lock(mx);
+
+        // If no in-flight requests, stop the background thread
+        if (inFlightRequests.size() == 0) {
+            SPDLOG_INFO("No requests to check");
+            return true;
+        }
+
+        // For each in-flight request that has opted in to be migrated,
+        // check if there is an opportunity to migrate
+        for (const auto& app : inFlightRequests) {
+            auto req = app.second.first;
+            auto originalDecision = *app.second.second;
+
+            // If we have already recorded a pending migration for this req,
+            // skip
+            if (canAppBeMigrated(originalDecision.appId) != nullptr) {
+                SPDLOG_INFO("Skipping app {} as migration opportunity has "
+                            "already been recorded",
+                            originalDecision.appId);
+                continue;
+            }
+
+            faabric::PendingMigrations msg;
+            msg.set_appid(originalDecision.appId);
+            // TODO - generate a new groupId here for processes to wait on
+            // during the migration? msg.set_groupid();
+
+            if (migrationStrategy ==
+                faabric::util::MigrationStrategy::BIN_PACK) {
```

Comment:
Note that this is quite similar to the `makeSchedulingDecision` code but with some differences that made it preferrable to be implemented as a separate bit of code:
* First, we are not actually scheduling new functions, we are just checking if a particular kind of re-arrangement is possible.
* Second, we don't enforce the changes, therefore we actually don't have to persist them (in the `HostResources`).